cordova-plugin-outline/android/java/org/outline/OutlinePlugin.java

```diff
@@ -300,29 +342,54 @@ public void onActivityResult(int request, int result, Intent data) {
 
   private void startVpnTunnel() {
     LOG.info("Starting VPN Tunnel");
-    if (vpnTunnelService == null) {
-      onVpnTunnelServiceNotBound(Action.START, startRequestTunnelId);
-      return;
-    }
-    vpnTunnelService.startTunnel(startRequestTunnelId, startRequestConfig);
+    final Bundle data = new Bundle();
+    data.putString(MessageData.TUNNEL_CONFIG.value, startRequestConfig.toString());
+    sendVpnServiceMessage(Action.START, startRequestTunnelId, data);
   }
 
   private void stopVpnTunnel(final String tunnelId) {
     LOG.info("Stopping VPN tunnel.");
-    if (vpnTunnelService == null) {
-      onVpnTunnelServiceNotBound(Action.STOP, tunnelId);
-      return;
-    }
-    vpnTunnelService.stopTunnel(tunnelId);
+    sendVpnServiceMessage(Action.STOP, tunnelId, null);
   }
 
   // Returns whether the VPN service is running a particular tunnel instance.
-  private boolean isTunnelActive(final String tunnelId) {
-    if (vpnTunnelService == null) {
-      onVpnTunnelServiceNotBound(Action.IS_RUNNING, tunnelId);
-      return false;
+  private void isTunnelActive(final String tunnelId) {
+    sendVpnServiceMessage(Action.IS_RUNNING, tunnelId, null);
+  }
+
+  // Initializes the error reporting framework on the VPN process.
+  private void initVpnErrorReporting(final String apiKey) {
+    Bundle data = new Bundle();
+    data.putString(MessageData.ERROR_REPORTING_API_KEY.value, apiKey);
+    sendVpnServiceMessage(Action.INIT_ERROR_REPORTING, null, data);
+  }
+
+  // Sends a message to the VPN service through its messenger. The VPN service must be bound.
+  void sendVpnServiceMessage(
```

Comment:
It seems like this is implementing an RPC mechanism over one-way `Message`s, using the "action" and "tunnelId" as the key to pair requests and responses.  It's hard to understand this in the context of this file, which _also_ handles the Cordova RPC.  I think we should probably do one or more of the following things:
* Use `IBinder.transact()` and `Binder.onTransact()` to get request-response pairing.
* Pass `CallbackContext.getCallbackId()` to the extension and back, so we can reconstruct the callback instead of caching it.
* Move the cross-process RPC into a separate file from the Cordova RPC.
* Define a generic cross-process RPC abstraction that we can use here.

Reply:
Based on your suggestions I re-implemented the VPN service process RPC through AIDL. 

- Calls to the VPN process are syncrhonous, simplifying the communication between Cordova JS and native. We no longer maintain a map of JS callbacks by tunnel ID/action pair. Instead, we have a map of tunnel status callbacks, which rely on asyncrhonous broadcasts.
- AIDL types reduce the complexity of unstructured message bundles.
- Added synchronization to the public VPN service methods to make AIDL implementation thread-safe.